File: validation/qos/durability/rclpy/qos_durability_pub.py
#!/usr/bin/env python3

# Copyright 2026 Int2DDS Project
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

import argparse
import sys
import time

import rclpy
from rclpy.executors import SingleThreadedExecutor
from rclpy.node import Node
try:
    from rclpy.event_handler import PublisherEventCallbacks
except ModuleNotFoundError:
    from rclpy.qos_event import PublisherEventCallbacks
from rclpy.qos import DurabilityPolicy, HistoryPolicy, QoSProfile, ReliabilityPolicy
from std_msgs.msg import String


def parse_args() -> argparse.Namespace:
    parser = argparse.ArgumentParser()
    parser.add_argument('--topic', default='qos_durability_topic')
    parser.add_argument('--durability', choices=['volatile', 'transient_local'], required=True)
    parser.add_argument('--reliability', choices=['reliable', 'best_effort'], default='reliable')
    parser.add_argument('--seed', required=True)
    parser.add_argument('--seed-delay', type=float, default=0.0)
    parser.add_argument('--keep-alive', type=float, default=5.0)
    parser.add_argument('--repeat-count', type=int, default=1)
    parser.add_argument('--repeat-period', type=float, default=0.5)
    parser.add_argument('--wait-for-subscriber-count', type=int, default=0)
    parser.add_argument('--wait-timeout', type=float, default=5.0)
    parser.add_argument('--enable-incompatible-events', action='store_true')
    return parser.parse_args()


def durability_policy(name: str) -> DurabilityPolicy:
    return (
        DurabilityPolicy.TRANSIENT_LOCAL
        if name == 'transient_local'
        else DurabilityPolicy.VOLATILE
    )


def reliability_policy(name: str) -> ReliabilityPolicy:
    return (
        ReliabilityPolicy.RELIABLE
        if name == 'reliable'
        else ReliabilityPolicy.BEST_EFFORT
    )


class DurabilityPublisher(Node):
    def __init__(self, args: argparse.Namespace) -> None:
        super().__init__('qos_durability_pub', start_parameter_services=False, enable_rosout=False)
        self.args = args
        self.done = False
        self.incompatible_events = 0
        self.start_time = time.time()
        self.publish_count = 0

        qos = QoSProfile(
            history=HistoryPolicy.KEEP_LAST,
            depth=1,
            reliability=reliability_policy(args.reliability),
            durability=durability_policy(args.durability),
        )
        publisher_kwargs = {}
        if args.enable_incompatible_events:
            publisher_kwargs['event_callbacks'] = PublisherEventCallbacks(
                incompatible_qos=self._incompatible_qos
            )
        self.publisher = self.create_publisher(String, args.topic, qos, **publisher_kwargs)
        self.observation_timer = self.create_timer(0.5, self._observe)
        self.publish_timer = self.create_timer(max(args.seed_delay, 0.01), self._publish_once)
        self.shutdown_timer = self.create_timer(
            max(args.seed_delay + args.keep_alive, 0.1),
            self._finish,
        )
        self.get_logger().info(
            'publisher ready: '
            f'topic={args.topic}, durability={args.durability}, reliability={args.reliability}'
        )

    def _publish_once(self) -> None:
        if self.args.wait_for_subscriber_count > 0:
            current_count = self.count_subscribers(self.args.topic)
            if current_count < self.args.wait_for_subscriber_count:
                if time.time() - self.start_time > self.args.wait_timeout:
                    self.publish_timer.cancel()
                    self.observation_timer.cancel()
                    self.shutdown_timer.cancel()
                    self.done = True
                    self.get_logger().error(
                        'publisher wait-for-match failed: '
                        f'expected={self.args.wait_for_subscriber_count}, current={current_count}'
                    )
                return
        msg = String()
        msg.data = self.args.seed
        self.publisher.publish(msg)
        self.publish_count += 1
        self.get_logger().info(f'published seed #{self.publish_count}: {msg.data}')
        if self.publish_count >= self.args.repeat_count:
            self.publish_timer.cancel()
        else:
            self.publish_timer.timer_period_ns = int(self.args.repeat_period * 1e9)

    def _incompatible_qos(self, event) -> None:
        self.incompatible_events += 1
        self.get_logger().info(
            'publisher incompatible qos event: '
            f'total_count={event.total_count}, total_count_change={event.total_count_change}, '
            f'last_policy_kind={event.last_policy_kind}'
        )

    def _observe(self) -> None:
        self.get_logger().info(
            f'subscriber_count observation: {self.count_subscribers(self.args.topic)}'
        )

    def _finish(self) -> None:
        self.shutdown_timer.cancel()
        self.observation_timer.cancel()
        self.done = True
        self.get_logger().info('publisher keep-alive finished')


def main() -> int:
    args = parse_args()
    rclpy.init()
    node = DurabilityPublisher(args)
    executor = SingleThreadedExecutor()
    executor.add_node(node)

    try:
        while rclpy.ok() and not node.done:
            executor.spin_once(timeout_sec=0.1)
    finally:
        executor.remove_node(node)
        node.destroy_node()
        executor.shutdown(timeout_sec=1.0)
        rclpy.shutdown()

    return 0


if __name__ == '__main__':
    sys.exit(main())

File: validation/qos/durability/rclcpp/qos_durability_pub.cpp
// Copyright 2026 Int2DDS Project
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

#include <chrono>
#include <memory>
#include <string>

#include "rclcpp/rclcpp.hpp"
#include "rclcpp/publisher_options.hpp"
#include "std_msgs/msg/string.hpp"

using namespace std::chrono_literals;

namespace
{

rclcpp::DurabilityPolicy parse_durability(const std::string & value)
{
  if (value == "transient_local") {
    return rclcpp::DurabilityPolicy::TransientLocal;
  }
  return rclcpp::DurabilityPolicy::Volatile;
}

rclcpp::ReliabilityPolicy parse_reliability(const std::string & value)
{
  if (value == "best_effort") {
    return rclcpp::ReliabilityPolicy::BestEffort;
  }
  return rclcpp::ReliabilityPolicy::Reliable;
}

class DurabilityPublisherCpp : public rclcpp::Node
{
public:
  DurabilityPublisherCpp(
    const std::string & topic,
    const std::string & durability,
    const std::string & reliability,
    const std::string & seed,
    double seed_delay_sec,
    double keep_alive_sec,
    int repeat_count,
    double repeat_period_sec,
    int wait_for_subscription_count,
    double wait_timeout_sec,
    bool enable_incompatible_events)
  : Node("qos_durability_pub_cpp",
      rclcpp::NodeOptions().start_parameter_services(false).enable_rosout(false)),
    topic_(topic),
    seed_(seed),
    keep_alive_sec_(keep_alive_sec),
    repeat_count_(repeat_count),
    repeat_period_sec_(repeat_period_sec),
    wait_for_subscription_count_(wait_for_subscription_count),
    wait_timeout_sec_(wait_timeout_sec)
  {
    rclcpp::QoS qos(rclcpp::KeepLast(1));
    qos.durability(parse_durability(durability));
    qos.reliability(parse_reliability(reliability));

    rclcpp::PublisherOptions options;
    if (enable_incompatible_events) {
      options.event_callbacks.incompatible_qos_callback =
        [this](rclcpp::QOSOfferedIncompatibleQoSInfo & info) {
          incompatible_events_++;
          RCLCPP_INFO(
            get_logger(),
            "publisher incompatible qos event: total_count=%d, total_count_change=%d,"
            "last_policy_kind=%d",
            info.total_count,
            info.total_count_change,
            info.last_policy_kind);
        };
    }

    publisher_ = create_publisher<std_msgs::msg::String>(topic, qos, options);
    observation_timer_ = create_wall_timer(500ms, [this]() {
          RCLCPP_INFO(get_logger(), "subscriber_count observation: %zu", count_subscribers(topic_));
    });
    publish_timer_ = create_wall_timer(
      std::chrono::duration_cast<std::chrono::nanoseconds>(std::chrono::duration<double>(
          std::max(seed_delay_sec, 0.01))),
      [this]() {publish_once();});
    shutdown_timer_ = create_wall_timer(
      std::chrono::duration_cast<std::chrono::nanoseconds>(std::chrono::duration<double>(
          std::max(seed_delay_sec + keep_alive_sec_, 0.1))),
      [this]() {
        shutdown_timer_->cancel();
        observation_timer_->cancel();
        done_ = true;
        RCLCPP_INFO(get_logger(), "publisher keep-alive finished");
      });

    start_time_ = now();
    RCLCPP_INFO(
      get_logger(),
      "publisher ready: topic=%s, durability=%s, reliability=%s",
      topic.c_str(),
      durability.c_str(),
      reliability.c_str());
  }

  bool done() const
  {
    return done_;
  }

private:
  void publish_once()
  {
    if (wait_for_subscription_count_ > 0) {
      const auto current_count = count_subscribers(topic_);
      if (static_cast<int>(current_count) < wait_for_subscription_count_) {
        const auto elapsed = (now() - start_time_).seconds();
        if (elapsed > wait_timeout_sec_) {
          publish_timer_->cancel();
          observation_timer_->cancel();
          shutdown_timer_->cancel();
          done_ = true;
          RCLCPP_ERROR(
            get_logger(),
            "publisher wait-for-match failed: expected=%d, current=%zu",
            wait_for_subscription_count_,
            current_count);
        }
        return;
      }
    }

    std_msgs::msg::String msg;
    msg.data = seed_;
    publisher_->publish(msg);
    publish_count_++;
    RCLCPP_INFO(get_logger(), "published seed #%d: %s", publish_count_, seed_.c_str());

    if (publish_count_ >= repeat_count_) {
      publish_timer_->cancel();
    } else {
      publish_timer_->reset();
      publish_timer_->cancel();
      publish_timer_ = create_wall_timer(
        std::chrono::duration_cast<std::chrono::nanoseconds>(std::chrono::duration<double>(
            repeat_period_sec_)),
        [this]() {publish_once();});
    }
  }

  std::string topic_;
  std::string seed_;
  double keep_alive_sec_;
  int repeat_count_;
  double repeat_period_sec_;
  int wait_for_subscription_count_;
  double wait_timeout_sec_;
  bool done_{false};
  int publish_count_{0};
  int incompatible_events_{0};
  rclcpp::Time start_time_;
  rclcpp::Publisher<std_msgs::msg::String>::SharedPtr publisher_;
  rclcpp::TimerBase::SharedPtr observation_timer_;
  rclcpp::TimerBase::SharedPtr publish_timer_;
  rclcpp::TimerBase::SharedPtr shutdown_timer_;
};

}  // namespace

int main(int argc, char ** argv)
{
  std::string topic = "qos_durability_topic";
  std::string durability;
  std::string reliability = "reliable";
  std::string seed;
  double seed_delay_sec = 0.0;
  double keep_alive_sec = 5.0;
  int repeat_count = 1;
  double repeat_period_sec = 0.5;
  int wait_for_subscription_count = 0;
  double wait_timeout_sec = 5.0;
  bool enable_incompatible_events = false;

  for (int i = 1; i < argc; ++i) {
    const std::string arg(argv[i]);
    auto next_value = [&](const char * name) -> std::string {
        if (i + 1 >= argc) {
          throw std::runtime_error(std::string("missing value for ") + name);
        }
        return argv[++i];
      };
    if (arg == "--topic") {
      topic = next_value("--topic");
    } else if (arg == "--durability") {
      durability = next_value("--durability");
    } else if (arg == "--reliability") {
      reliability = next_value("--reliability");
    } else if (arg == "--seed") {
      seed = next_value("--seed");
    } else if (arg == "--seed-delay") {
      seed_delay_sec = std::stod(next_value("--seed-delay"));
    } else if (arg == "--keep-alive") {
      keep_alive_sec = std::stod(next_value("--keep-alive"));
    } else if (arg == "--repeat-count") {
      repeat_count = std::stoi(next_value("--repeat-count"));
    } else if (arg == "--repeat-period") {
      repeat_period_sec = std::stod(next_value("--repeat-period"));
    } else if (arg == "--wait-for-subscriber-count") {
      wait_for_subscription_count = std::stoi(next_value("--wait-for-subscriber-count"));
    } else if (arg == "--wait-timeout") {
      wait_timeout_sec = std::stod(next_value("--wait-timeout"));
    } else if (arg == "--enable-incompatible-events") {
      enable_incompatible_events = true;
    } else {
      throw std::runtime_error("unknown argument: " + arg);
    }
  }

  if (durability.empty()) {
    throw std::runtime_error("--durability is required");
  }
  if (seed.empty()) {
    throw std::runtime_error("--seed is required");
  }

  rclcpp::init(argc, argv);
  auto node = std::make_shared<DurabilityPublisherCpp>(
    topic,
    durability,
    reliability,
    seed,
    seed_delay_sec,
    keep_alive_sec,
    repeat_count,
    repeat_period_sec,
    wait_for_subscription_count,
    wait_timeout_sec,
    enable_incompatible_events);

  rclcpp::executors::SingleThreadedExecutor executor;
  executor.add_node(node);
  while (rclcpp::ok() && !node->done()) {
    executor.spin_once(100ms);
  }
  executor.remove_node(node);
  rclcpp::shutdown();
  return 0;
}
